sound/soc/codecs/pcm512x.c

```diff
@@ -1503,6 +1506,121 @@ const struct regmap_config pcm512x_regmap = {
 };
 EXPORT_SYMBOL_GPL(pcm512x_regmap);
 
+static int pcm512x_gpio_get_direction(struct gpio_chip *chip,
+				      unsigned int offset)
+{
+	struct pcm512x_priv *pcm512x = gpiochip_get_data(chip);
+	unsigned int val;
+	int ret;
+
+	ret = regmap_read(pcm512x->regmap, PCM512x_GPIO_EN, &val);
+	if (ret < 0)
+		return ret;
+
+	val = (val >> offset) & 1;
+
+	/* val is 0 for input, 1 for output, return inverted */
+	return !val;
+}
+
+static int pcm512x_gpio_direction_input(struct gpio_chip *chip,
+					unsigned int offset)
+{
+	struct pcm512x_priv *pcm512x = gpiochip_get_data(chip);
+	unsigned int mask;
+
+	mask = 1 << offset;
+
+	regmap_update_bits(pcm512x->regmap, PCM512x_GPIO_EN,
+			   mask, 0);
+	return 0;
+}
+
+static int pcm512x_gpio_direction_output(struct gpio_chip *chip,
+					 unsigned int offset,
+					 int value)
+{
+	struct pcm512x_priv *pcm512x = gpiochip_get_data(chip);
+	unsigned int mask;
+	unsigned int bits;
+	unsigned int reg;
+
+	mask = 1 << offset;
+	bits = value << offset;
+
+	/* select Register GPIOx output for OUTPUT_x (1..6) */
+	reg = PCM512x_GPIO_OUTPUT_1 + offset;
+	regmap_update_bits(pcm512x->regmap, reg, 0x0f, 0x02);
+
+	/* enable output x */
+	regmap_update_bits(pcm512x->regmap, PCM512x_GPIO_EN,
+			   mask, mask);
+
+	/* set value */
+	regmap_update_bits(pcm512x->regmap, PCM512x_GPIO_CONTROL_1,
+			   mask, bits);
```

Comment:
Just a note: Looking at ASoC codecs, majority don't have error handling for regmap updates (but some do), so established practise seems mixed.  But based on the interface, should be checked..

Reply:
I'll add error-checks. this code was copy/pasted from somewhere else, but it's not because they didnt' add error check that we shouldn't. register access can fail on i2C, it's not safe to presume things are 100% reliable.